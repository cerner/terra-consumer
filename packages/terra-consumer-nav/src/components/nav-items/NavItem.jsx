import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import DangerousHtml from '../dangerous-html/DangerousHtml';
import styles from './NavItem.scss';
import NavToggler from '../nav-toggler/NavToggler';
import SmartLink from '../smart-link/SmartLink';
import navItemShape from '../../NavPropShapes';

const cx = classNames.bind(styles);

const propTypes = {
  ...navItemShape,
  /**
   * The id of the toggle component to be toggled.
   */
  toggleId: PropTypes.number,
  /**
   * Function callback when the toggle component is clicked.
   */
  handleToggle: PropTypes.func.isRequired,
  /**
   * Whether or not the content should be toggled open or not.
   */
  isOpen: PropTypes.bool,
  /**
   * Function to be applied to the generated link.
   */
  handleClick: PropTypes.func,
  /**
   * Items to be displayed within the toggler.
   */
  children: PropTypes.node,
};

const defaultProps = {
  isActive: false,
  handleToggle: () => {},
};

const NavItem = ({
  url,
  text,
  icon,
  target,
  isActive,
  badgeValue,
  toggleId,
  handleToggle,
  isOpen,
  isExternal,
  handleClick,
  children,
  ...customProps
}) => {
  const activeClass = cx('active');

  const itemLabel = (<div>
    <DangerousHtml text={text} />
    { badgeValue > 0 &&
      <div className={cx('badge')}>{badgeValue}</div>
    }
  </div>);

  const itemText = (
    <div className={cx('item')}>
      { icon && <div className={cx('icon')}>{icon}</div> }
      <div className={cx('label')}>{itemLabel}</div>
    </div>
  );
  return (children && children.length > 0) ?
    (<NavToggler
      {...customProps}
      header={itemText}
      handleToggle={open => handleToggle(toggleId, open)}
      isOpen={isOpen}
      className={cx('nested', customProps.className)}
    >
      {children}
    </NavToggler>)
    :
    (<div className={isActive ? activeClass : ''}>
      <SmartLink
        {...customProps}
        url={url}
        target={target}
        isExternal={isExternal}
        activeClass={activeClass}
        handleClick={handleClick}
        className={cx('link', customProps.className)}
      >
        {itemText}
      </SmartLink>
    </div>);
};

NavItem.propTypes = propTypes;
NavItem.defaultProps = defaultProps;

export default NavItem;
