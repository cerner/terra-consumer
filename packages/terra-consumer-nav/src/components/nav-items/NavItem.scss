@import '../../variables';

:local {
  .item {
    color: var(--terra-consumer-nav-item-color, #fff);
    display: flex;
    font-size: 16px;
    line-height: 20px;
    padding: 7px 0 7px 25px;
    position: relative;
    text-decoration: none;
  }

  .icon {
    padding-right: 20px;
  }

  .link {
    display: block;
    text-decoration: none;

    &:active {}

    &:hover,
    &:focus {
      background: rgba(255, 255, 255, 0.4);
      color: var(--terra-consumer-nav-item-color);
      text-decoration: underline;
    }
  }

  /* stylelint-disable selector-max-compound-selectors, max-nesting-depth */
  .nested {
    .link {
      .item {
        padding-left: 60px;
      }

      .icon {
        padding-right: 10px;
      }
    }
  }

  .badge {
    align-self: center;
    background: var(--terra-consumer-badge-background, #fff);
    border-radius: 10px;
    color: var(--terra-consumer-badge-color, #3d3d3d);
    display: inline-block;
    font-size: 12px;
    margin: 0 10px;
    padding: 0 10px;
  }

  .active {
    background: var(--terra-consumer-nav-active-item-background, rgba(255, 255, 255, 0.5));
    background-attachment: fixed;
    background-size: cover;
    font-weight: 900;
    position: relative;

    &::before {
      background: var(--terra-consumer-nav-item-color, #383f42);
      border-radius: 3px;
      bottom: -0;
      content: '';
      left: 0;
      position: absolute;
      top: -0;
      width: 4px;
      @media screen and (max-width: $terra-medium-breakpoint) {
        background: #fff;
      }
    }
  }
}
