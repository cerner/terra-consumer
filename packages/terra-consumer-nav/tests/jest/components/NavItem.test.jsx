import React from 'react';
import NavItem from '../../../src/components/nav-items/NavItem';

const testData = {
  url: '#testPath',
  text: 'testIcon',
  isActive: false,
};

const toggleData = {
  toggleId: 1,
  handleToggle: () => {},
  isOpen: false,
  children: <div>test</div>,
};

describe('Nav Item', () => {
  // Snapshot Tests
  it('should render a default component', () => {
    const wrapper = shallow(<NavItem {...testData} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should apply custom classes', () => {
    const wrapper = shallow(<NavItem {...testData} className="test-class" />);
    expect(wrapper).toMatchSnapshot();
  });

  describe('badge', () => {
    it('should render when a number > 0 is provided', () => {
      const wrapper = shallow(<NavItem {...testData} badgeValue={1} />);
      expect(wrapper).toMatchSnapshot();
    });

    it('should render when a string is provided', () => {
      const wrapper = shallow(<NavItem {...testData} badgeValue="!" />);
      expect(wrapper).toMatchSnapshot();
    });
  });
});

describe('Nav Item as Toggler', () => {
  // Snapshot Tests
  it('should render a nav as a toggle component', () => {
    const wrapper = shallow(<NavItem {...testData} {...toggleData} />);
    expect(wrapper).toMatchSnapshot();
  });
});
