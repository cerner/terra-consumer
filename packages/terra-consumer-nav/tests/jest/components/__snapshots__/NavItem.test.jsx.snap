// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nav Item as Toggler should render a nav as a toggle component 1`] = `
<div
  className=""
>
  <SmartLink
    activeClass="active"
    className="link"
    handleClick={[Function]}
    isExternal={false}
    target="_self"
    url="#testPath"
  >
    <div
      className="item"
    >
      <div
        className="label"
      >
        <div>
          <DangerousHtml
            text="testIcon"
          />
        </div>
      </div>
    </div>
  </SmartLink>
</div>
`;

exports[`Nav Item should apply custom classes 1`] = `
<div
  className=""
>
  <SmartLink
    activeClass="active"
    className="link test-class"
    handleClick={[Function]}
    isExternal={false}
    target="_self"
    url="#testPath"
  >
    <div
      className="item"
    >
      <div
        className="label"
      >
        <div>
          <DangerousHtml
            text="testIcon"
          />
        </div>
      </div>
    </div>
  </SmartLink>
</div>
`;

exports[`Nav Item should render a default component 1`] = `
<div
  className=""
>
  <SmartLink
    activeClass="active"
    className="link"
    handleClick={[Function]}
    isExternal={false}
    target="_self"
    url="#testPath"
  >
    <div
      className="item"
    >
      <div
        className="label"
      >
        <div>
          <DangerousHtml
            text="testIcon"
          />
        </div>
      </div>
    </div>
  </SmartLink>
</div>
`;
